be/src/storage/binlog_manager.cpp

```diff
@@ -114,42 +115,218 @@ void BinlogManager::_apply_build_result(BinlogBuildResult* result) {
+        _total_wait_reader_binlog_file_size -= file_meta->file_size();
+        for (int64_t rowset_id : file_meta->rowsets()) {
+            int32_t count = --_wait_reader_rowset_count_map[rowset_id];
+            if (count == 0) {
+                _wait_reader_rowset_count_map.erase(rowset_id);
+                _total_wait_reader_rowset_data_size -= _rowset_fetcher->get_rowset(rowset_id)->data_disk_size();
+            }
+        }
+        _unused_binlog_file_ids.blocking_put(file_meta->id());
+        if (first_file_id == -1) {
+            first_file_id = file_meta->id();
+        }
+        last_file_id = file_meta->id();
+        num_files += 1;
+
+        _wait_reader_binlog_files.pop_front();
+
+        VLOG(3) << "Binlog file finished to wait readers, tablet: " << _tablet_id << ", file id: " << file_meta->id();
+    }
+    LOG(INFO) << "Check wait reader binlog files, tablet: " << _tablet_id << ", num unused files: " << num_files
+              << ", first_file_id: " << first_file_id << ", last_file_id: " << last_file_id
+              << ", num of still wait files: " << _wait_reader_binlog_files.size();
+}
+
+void BinlogManager::_check_alive_binlog_files(int64_t current_second, int64_t binlog_ttl_second,
+                                              int64_t binlog_max_size) {
+    int64_t expiration_time = current_second - binlog_ttl_second;
+    int64_t active_file_id = _active_binlog_writer != nullptr ? _active_binlog_writer->file_id() : -1;
+    // binlog files should be deleted in the order by the file id, and if there are previous
+    // binlog files waiting for readers, the latter should also wait for the readers
+    bool need_wait_reader = !_wait_reader_binlog_files.empty();
+    int64_t first_file_id = -1;
+    int64_t last_file_id = -1;
+    int64_t num_files = 0;
+    int64_t num_unused_files = 0;
+    for (auto it = _alive_binlog_files.begin(); it != _alive_binlog_files.end();) {
+        auto& binlog_file = it->second;
+        auto& meta = binlog_file->file_meta();
+        // skip to clean up the active file
+        if (meta->id() == active_file_id) {
+            break;
+        }
+        bool expired = meta->end_timestamp_in_us() / 1000000 < expiration_time;
+        bool overcapacity = _total_alive_binlog_file_size + _total_alive_rowset_data_size > binlog_max_size;
+        if (!expired && !overcapacity) {
+            break;
+        }
+
+        if (binlog_file->reader_count() > 0) {
+            need_wait_reader = true;
+        }
+
+        _total_alive_binlog_file_size -= meta->file_size();
+        if (need_wait_reader) {
+            _wait_reader_binlog_files.push_back(binlog_file);
+            _total_wait_reader_binlog_file_size += meta->file_size();
+        } else {
+            _unused_binlog_file_ids.blocking_put(meta->id());
+        }
+
+        for (int64_t rowset_id : meta->rowsets()) {
+            int32_t count = --_alive_rowset_count_map[rowset_id];
```

Comment:
add DCHECK or some code to prevent _alive_rowset_count_map[rowset_id] not found if bug exists

Reply:
Fixed